routemaster/cron.py

```diff
@@ -18,102 +20,138 @@
     MetadataTrigger,
 )
 from routemaster.state_machine import (
-    IsExitingCheck,
-    StateProcessor,
-    process_gate_trigger,
-    process_action_retries,
-    process_gate_metadata_retries,
+    LabelStateProcessor,
+    process_cron,
+    process_gate,
+    process_action,
+    labels_in_state,
+    labels_needing_metadata_update_retry_in_gate,
 )
 
 logger = logging.getLogger(__name__)
 
+IsTerminating = Callable[[], bool]
+LabelProvider = Callable[[StateMachine, State, Any], Iterable[str]]
 
-class _Process(NamedTuple):
-    fn: StateProcessor
-    state: State
-    app: App
-    state_machine: StateMachine
-    is_terminating: IsExitingCheck
+CronProcessor = Callable[
+    [LabelStateProcessor, LabelProvider, State, StateMachine],
+    None,
+]
 
-    def __call__(self):
-        name = getattr(self.fn, '__name__', 'process')
-        logger.info(
-            f"Started cron {name} for state {self.state.name} in "
-            f"{self.state_machine.name}",
-        )
 
-        try:
-            time_start = time.time()
-            self.fn(
-                self.app,
-                self.state,
-                self.state_machine,
-                self.is_terminating,
-            )
-            duration = time.time() - time_start
-        except Exception as e:
-            print(e)
+# The cron configuration works by building up a partially applied function
+# `process_job`, at multiple levels. This allows us to hook an intermediate
+# stage for testing.
+
+def process_job(
+    *,
+    # Bound at the cron thread level
+    app: App,
+    is_terminating: IsTerminating,
```

Comment:
Building on our previous conversation about deployment/clean-shutdown etc, I wonder whether (in the future) hanging `is_terminating` off the `app` would be the way to go?

Definitely something for the future though.